src/pudl_archiver/utils.py

```diff
@@ -74,3 +77,56 @@ def add_to_archive_stable_hash(archive: zipfile.ZipFile, filename, data: bytes):
     # also this can't be set in the constructor as of 2024-02-09
     info.compress_type = zipfile.ZIP_DEFLATED
     archive.writestr(info, data)
+
+
+async def _rate_limited_scheduler(
+    tasks: list[typing.Awaitable],
+    result_queue: asyncio.Queue,
+    rate_limit: int = 10,
+):
+    """Launch rate limited tasks."""
+
+    async def _run_task(task: typing.Awaitable, result_queue: asyncio.Queue):
+        await result_queue.put(await task)
+
+    scheduled_tasks = []
+    for task in tasks:
+        scheduled_tasks.append(asyncio.create_task(_run_task(task, result_queue)))
+        await asyncio.sleep(1 / rate_limit)
+
+    # Wait for all tasks to complete
+    await asyncio.gather(*scheduled_tasks)
+    await result_queue.put("tasks_complete")
+
+
+async def rate_limit_tasks(tasks: list[typing.Awaitable], rate_limit: int = 10):
```

Comment:
I added this function for the sec archiver, but decided to put it in this infrastructure PR. It just launches download tasks at a specified rate until they're all done, and returns outputs as tasks complete.